synapse/handlers/_base.py

```diff
@@ -93,20 +101,32 @@ def allowed(event, user_id, is_peeking):
             else:
                 membership = None
 
+            # if the user was a member of the room at the time of the event,
+            # they can see it.
             if membership == Membership.JOIN:
                 return True
 
             if event.type == EventTypes.RoomHistoryVisibility:
-                return not is_peeking
+                # XXX why are m.room.history_visibility events special?
+                # return True
+                pass
 
             if visibility == "shared":
-                return True
-            elif visibility == "joined":
-                return membership == Membership.JOIN
+                # user can also see the event if he has become a member since
+                # the event
+                #
+                # XXX: if the user has subsequently joined and then left again,
+                # ideally we would share history up to the point they left. But
+                # we don't know when they left.
+                return not is_peeking
             elif visibility == "invited":
+                # user can also see the event if he was *invited* at the time
+                # of the event.
                 return membership == Membership.INVITE
 
-            return True
+            # presumably visibility is "joined"; we weren't a member at the
+            # time of the event, so we're done.
```

Comment:
I think the server is supposed to assume that the room is "shared" if it doesn't understand the value of history_visibility.

https://matrix.org/speculator/spec/HEAD/client_server.html#id31